### PR TITLE
Add Rust server and client for `get_simple` example

### PR DESCRIPTION
**Pull request by mbrobbel**
A simple synchronous Rust (`std`-based) server and client for the `get_simple` example.

This skips over all `http` request/response parsing and error handling, but the examples can be updated later to use [`hyper`](https://docs.rs/hyper/latest/hyper/) (when there is [async Arrow IPC support in arrow-rs](https://github.com/apache/arrow-rs/issues/1207)).

```
❯ cargo r -p server --release
    Finished release [optimized] target(s) in 0.04s
     Running `target/release/server`
2024-03-07T20:14:47.266470Z  INFO data{TOTAL_RECORDS=100000000 RECORDS_PER_BATCH=4096}: server: Generating random data
2024-03-07T20:14:48.109722Z  INFO data{TOTAL_RECORDS=100000000 RECORDS_PER_BATCH=4096}: server: close time.busy=843ms time.idle=1.12µs
2024-03-07T20:14:48.109820Z  INFO server: Listening bind_addr=127.0.0.1:8000
2024-03-07T20:15:17.403848Z  INFO Writing Arrow IPC stream{remote_peer=127.0.0.1:50291}: server: Incoming connection
2024-03-07T20:15:18.006722Z  INFO Writing Arrow IPC stream{remote_peer=127.0.0.1:50291}: server: close time.busy=603ms time.idle=2.08µs
```

```
❯ cargo r -p client --release
    Finished release [optimized] target(s) in 0.03s
     Running `target/release/client`
2024-03-07T20:15:17.403862Z  INFO Reading Arrow IPC stream{addr=127.0.0.1:8000}: client: Connected
2024-03-07T20:15:18.007890Z  INFO Reading Arrow IPC stream{addr=127.0.0.1:8000}: client: batches=24415 rows=100000000
2024-03-07T20:15:18.147404Z  INFO Reading Arrow IPC stream{addr=127.0.0.1:8000}: client: close time.busy=744ms time.idle=1.21µs
```

### CODE REVIEW

**Comment by ianmcook** [comment]
Thanks @mbrobbel !

Does the `Cargo.lock` file need to be committed here?

**Comment by mbrobbel** [comment]
> Does the `Cargo.lock` file need to be committed here?

It's the recommended approach:
- https://doc.rust-lang.org/stable/cargo/faq.html#why-have-cargolock-in-version-control
- https://blog.rust-lang.org/2023/08/29/committing-lockfiles.html

**Comment by ianmcook** [comment]
> It's the recommended approach

After reading the guidance on this from the Cargo team, I think I'm -1 on including it in this case. The project here is intended to serve as an example for other projects to build on, not as a standalone project where the top priority is for the binary to build without errors. It seems undesirable in this case to have 900+ extra lines just to ensure the latter. The spirit of this area of the repo is to keep everything very minimal and not have any unnecessary assets even if they are black boxes.

**Comment by ianmcook** [comment]
I tested this Rust server with the clients implemented in other languages. That worked fine.

But I ran into problems when I tested this Rust client with the servers implemented in other languages.

File: http/get_simple/rs/client/src/main.rs
Comment (ianmcook) [inline http/get_simple/rs/client/src/main.rs:21]: Can we make the client print elapsed time?

Feel free to ignore my suggestions here and do it differently :) Thanks
```suggestion
    net::{IpAddr, Ipv4Addr, SocketAddr, TcpStream},
    time::{Instant},
```

File: http/get_simple/rs/client/src/main.rs
Comment (ianmcook) [inline http/get_simple/rs/client/src/main.rs:31]:
```suggestion

    let start = Instant::now();

```

File: http/get_simple/rs/client/src/main.rs
Comment (ianmcook) [inline http/get_simple/rs/client/src/main.rs:64]:
```suggestion
    }
    let end = Instant::now();
    info!(
        elapsed = format!("{:.2} seconds", end.duration_since(start).as_secs_f64())
    );
```

**Comment by ianmcook** [comment]
The client works with the Python server example now, but I see errors when I try to use it with the Java and Go server examples.

With the Java server, I see this error:
```
thread 'main' panicked at client/src/main.rs:52:77:
called `Result::unwrap()` on an `Err` value: ParseError("Unable to get root as message: Unaligned { position: 3795177799, unaligned_type: \"i32\", error_trace: ErrorTrace([]) }")
```

With the Go server, I see this error:
```
thread 'main' panicked at library/alloc/src/raw_vec.rs:571:5:
capacity overflow
```
cc @zeroshade

File: http/get_simple/rs/client/src/main.rs
Comment (mbrobbel) [reply http/get_simple/rs/client/src/main.rs:21]: I added another span around the full example. With `.with_span_events(FmtSpan::CLOSE)` the `fmt` subscriber outputs the span duration:
```
2024-03-08T18:20:22.956542Z  INFO get_simple:Reading Arrow IPC stream{addr=127.0.0.1:8000}: client: Connected
2024-03-08T18:20:23.617011Z  INFO get_simple:Reading Arrow IPC stream{addr=127.0.0.1:8000}: client: batches=24415 rows=100000000
2024-03-08T18:20:23.748391Z  INFO get_simple:Reading Arrow IPC stream{addr=127.0.0.1:8000}: client: close time.busy=792ms time.idle=666ns
2024-03-08T18:20:23.748413Z  INFO get_simple: client: close time.busy=792ms time.idle=708ns
```

Note the `close time.busy=792ms time.idle=708ns` part.

**Comment by mbrobbel** [comment]
> The client works with the Python server example now, but I see errors when I try to use it with the Java and Go server examples.
> 
> With the Java server, I see this error:
> 
> ```
> thread 'main' panicked at client/src/main.rs:52:77:
> called `Result::unwrap()` on an `Err` value: ParseError("Unable to get root as message: Unaligned { position: 3795177799, unaligned_type: \"i32\", error_trace: ErrorTrace([]) }")
> ```
> 
> With the Go server, I see this error:
> 
> ```
> thread 'main' panicked at library/alloc/src/raw_vec.rs:571:5:
> capacity overflow
> ```
> 
> cc @zeroshade

The problem is caused by the `BufReader` in the client. In the case of the errors it buffered some bytes from the beginning of the body. I'll fix it.

**Comment by mbrobbel** [comment]
Do we want CI?

**Comment by ianmcook** [comment]
> Do we want CI?

Not yet. We will want to add CI later, once the examples like these become part of an official conventions document on the Arrow docs site. I think in general we want to keep this repo free of CI.

**Review by ianmcook — APPROVED** [review]
I tested these with all the other client and server examples and they work as expected. Thanks very much @mbrobbel!

@wjones127 or anyone else watching here: is there anything else you'd like to see here before we merge this?

**Comment by ianmcook** [comment]
@mbrobbel I'm curious how feasible it would be to make the client example handle HTTP/1.1 responses with [chunked transfer encoding](https://en.wikipedia.org/wiki/Chunked_transfer_encoding)? I think this would just involve handling lines with `\r\n` a bit differently.

**Review by ianmcook — APPROVED** [review]
Thanks @mbrobbel! I re-tested this new version that includes chunked encoding support against all the other server examples, toggling chunked encoding on and off on the servers, and it works as expected.

**Comment by ianmcook** [comment]
I will merge later today if there are no other comments.

File: http/get_simple/rs/client/src/main.rs
Comment (wjones127) [inline http/get_simple/rs/client/src/main.rs:61]: It feels a bit sad we are manually parsing headers and such. Makes the example a bit harder to follow.

**Review by wjones127 — APPROVED** [review]
Seems okay, though I would definitely prefer it written in something higher-level like hyper. As-is, I'm not sure how many developers would use this logic.

File: http/get_simple/rs/client/src/main.rs
Comment (mbrobbel) [reply http/get_simple/rs/client/src/main.rs:61]: 100% agreed. I started with `hyper` but moved to `std` because of https://github.com/apache/arrow-rs/issues/1207.

Comment (ianmcook) [reply http/get_simple/rs/client/src/main.rs:61]: Agreed. Of all the examples here, this is the only one that directly implements low-level HTTP details instead of depending on a library to do that.

But the main objective of these `get_simple` examples is only to ensure that the IPC implementations in the mainstream Arrow libraries can function and interoperate over HTTP. This achieves that objective.

File: http/get_simple/rs/client/README.md
Comment (ianmcook) [inline http/get_simple/rs/client/README.md:32]:
````suggestion
```
> [!NOTE]  
> This client example implements low-level HTTP/1.1 details directly, instead of using an HTTP library. We intend to update the example to use [hyper](https://docs.rs/hyper/latest/hyper/) after [arrow-rs has an async Arrow IPC reader](https://github.com/apache/arrow-rs/issues/1207)).
````

File: http/get_simple/rs/server/README.md
Comment (ianmcook) [inline http/get_simple/rs/server/README.md:32]:
````suggestion
```
> [!NOTE]  
> This server example implements low-level HTTP/1.1 details directly, instead of using an HTTP library. We intend to update the example to use [hyper](https://docs.rs/hyper/latest/hyper/) after [arrow-rs has an async Arrow IPC writer](https://github.com/apache/arrow-rs/issues/1207)).
````

**Comment by tustvold** [comment]
FYI with https://github.com/apache/arrow-rs/pull/5531 it should be possible to switch this to use hyper

**Pull request merged**
